File: labs/lab-03-cache-aside/ruby/solution.rb
# Lab 03: Cache-Aside with Postgres + Redis — Ruby reference solution.
#
#   Run:  ruby solution.rb
require_relative '../../../tooling/ruby/labkit/lib/labkit'

TTL = 60
def user_key(id) = "user:#{id}"

# Part 1 — cache-aside read.
def get_user_profile(user_id)
  key = user_key(user_id)
  cached = Labkit::CACHE.get_json(key)
  return cached if cached

  row = Labkit::DB.query_one('SELECT id, name, email, plan FROM users WHERE id = $1', [user_id])
  return nil unless row

  Labkit::CACHE.set_json(key, row, TTL)
  row
end

def update_user_plan(user_id, plan)
  Labkit::DB.exec('UPDATE users SET plan = $1, updated_at = now() WHERE id = $2', [plan, user_id])
  Labkit::CACHE.delete(user_key(user_id))
end

# Part 2 — single-flight: a per-key lock + double-check so only one thread
# queries Postgres on a cold hot-key miss.
LOCKS = {}
LOCKS_GUARD = Mutex.new

def key_lock(key)
  LOCKS_GUARD.synchronize { LOCKS[key] ||= Mutex.new }
end

def get_user_profile_singleflight(user_id)
  key = user_key(user_id)
  cached = Labkit::CACHE.get_json(key)
  return cached if cached

  key_lock(key).synchronize do
    cached = Labkit::CACHE.get_json(key) # double-check
    return cached if cached

    row = Labkit::DB.query_one('SELECT id, name, email, plan FROM users WHERE id = $1', [user_id])
    return nil unless row

    Labkit::CACHE.set_json(key, row, TTL)
    row
  end
end

def assert(cond, msg)
  raise "CHECK FAILED: #{msg}" unless cond
end

Labkit::DB.exec("UPDATE users SET plan = 'pro' WHERE id = 1")
Labkit::CACHE.flush
Labkit::DB.reset_counters

u = get_user_profile(1)
assert(u && u['id'].to_i == 1, 'user 1 should be fetched')
assert(Labkit::DB.query_count >= 1, 'first read should hit Postgres')
assert(Labkit::CACHE.exists?(user_key(1)), 'first read should populate the cache')

Labkit::DB.reset_counters
u2 = get_user_profile(1)
assert(u2['id'].to_i == 1, 'second read returns the user')
assert(Labkit::DB.query_count == 0, 'a cache hit must not query Postgres')

update_user_plan(1, 'enterprise')
assert(!Labkit::CACHE.exists?(user_key(1)), 'update must invalidate the cache')
assert(get_user_profile(1)['plan'] == 'enterprise', 'refilled with the new plan')

assert(get_user_profile(999_999).nil?, 'missing user returns nil')

Labkit::CACHE.flush
Labkit::DB.reset_counters
threads = Array.new(50) { Thread.new { get_user_profile_singleflight(1) } }
threads.each(&:join)
assert(Labkit::DB.query_count == 1, 'single-flight: only one DB query for 50 concurrent reads')

puts 'OK — cache-aside, invalidation, and single-flight (1 DB query for 50 concurrent reads)'

File: labs/lab-03-cache-aside/rust/solution.rs
// Lab 03: Cache-Aside with Postgres + Redis — Rust reference solution.
//
// labkit gives you ready db() and cache() handles — no setup.
//   Run:  cargo run --bin solution
use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};
use std::thread;

use serde_json::json;

const TTL: u64 = 60;

#[derive(Clone)]
struct User {
    id: i32,
    name: String,
    email: String,
    plan: String,
}

fn user_key(id: i32) -> String {
    format!("user:{id}")
}

fn user_from_row(row: &labkit::Row) -> User {
    User { id: row.get("id"), name: row.get("name"), email: row.get("email"), plan: row.get("plan") }
}

fn user_to_json(u: &User) -> String {
    json!({"id": u.id, "name": u.name, "email": u.email, "plan": u.plan}).to_string()
}

fn user_from_json(s: &str) -> User {
    let v: serde_json::Value = serde_json::from_str(s).unwrap();
    User {
        id: v["id"].as_i64().unwrap() as i32,
        name: v["name"].as_str().unwrap().to_string(),
        email: v["email"].as_str().unwrap().to_string(),
        plan: v["plan"].as_str().unwrap().to_string(),
    }
}

// Part 1 — cache-aside read.
fn get_user_profile(id: i32) -> Option<User> {
    let key = user_key(id);
    if let Some(s) = labkit::cache().get(&key) {
        return Some(user_from_json(&s));
    }
    let row = labkit::db().query_one("SELECT id, name, email, plan FROM users WHERE id = $1", &[&id])?;
    let u = user_from_row(&row);
    labkit::cache().set(&key, &user_to_json(&u), Some(TTL));
    Some(u)
}

fn update_user_plan(id: i32, plan: &str) {
    labkit::db().exec("UPDATE users SET plan = $1, updated_at = now() WHERE id = $2", &[&plan, &id]);
    labkit::cache().del(&user_key(id));
}

// Part 2 — single-flight (per-key lock + double-check).
fn locks() -> &'static Mutex<HashMap<String, Arc<Mutex<()>>>> {
    static L: OnceLock<Mutex<HashMap<String, Arc<Mutex<()>>>>> = OnceLock::new();
    L.get_or_init(|| Mutex::new(HashMap::new()))
}

fn key_lock(key: &str) -> Arc<Mutex<()>> {
    let mut m = locks().lock().unwrap();
    m.entry(key.to_string()).or_insert_with(|| Arc::new(Mutex::new(()))).clone()
}

fn get_user_profile_singleflight(id: i32) -> Option<User> {
    let key = user_key(id);
    if let Some(s) = labkit::cache().get(&key) {
        return Some(user_from_json(&s));
    }
    let lock = key_lock(&key);
    let _guard = lock.lock().unwrap();
    if let Some(s) = labkit::cache().get(&key) {
        return Some(user_from_json(&s)); // double-check
    }
    let row = labkit::db().query_one("SELECT id, name, email, plan FROM users WHERE id = $1", &[&id])?;
    let u = user_from_row(&row);
    labkit::cache().set(&key, &user_to_json(&u), Some(TTL));
    Some(u)
}

fn assert(cond: bool, msg: &str) {
    if !cond {
        panic!("CHECK FAILED: {msg}");
    }
}

fn main() {
    labkit::db().exec("UPDATE users SET plan = 'pro' WHERE id = 1", &[]);
    labkit::cache().flush();
    labkit::db().reset_counters();

    let u = get_user_profile(1);
    assert(u.as_ref().map(|x| x.id) == Some(1), "user 1 should be fetched");
    assert(labkit::db().query_count() >= 1, "first read should hit Postgres");
    assert(labkit::cache().exists(&user_key(1)), "first read should populate the cache");

    labkit::db().reset_counters();
    let u2 = get_user_profile(1).unwrap();
    assert(u2.id == 1, "second read returns the user");
    assert(labkit::db().query_count() == 0, "a cache hit must not query Postgres");

    update_user_plan(1, "enterprise");
    assert(!labkit::cache().exists(&user_key(1)), "update must invalidate the cache");
    assert(get_user_profile(1).unwrap().plan == "enterprise", "refilled with the new plan");

    assert(get_user_profile(999_999).is_none(), "missing user returns None");

    labkit::cache().flush();
    labkit::db().reset_counters();
    let handles: Vec<_> = (0..50).map(|_| thread::spawn(|| get_user_profile_singleflight(1))).collect();
    for h in handles {
        h.join().unwrap();
    }
    assert(labkit::db().query_count() == 1, "single-flight: only one DB query for 50 concurrent reads");

    println!("OK — cache-aside, invalidation, and single-flight (1 DB query for 50 concurrent reads)");
}
